orttraining/orttraining/training_api/module.cc

```diff
@@ -48,62 +49,103 @@ Status Parameter::ResetGrad() {
-      weights_.push_back(it->second->Data());
     } else if (input_name == ACCUMULATE_GRAD_CONTROL_INPUT_NAME) {
       reset_grad_name.emplace_back(input_name);
     } else if (utils::GetParamNameFromGradient(input_name, param_name)) {
+      param_name_to_grad_input_index_map.insert({param_name, grad_input_names.size()});
       grad_input_names.emplace_back(input_name);
-      // create gradient buffer
-      auto it = named_parameters_.find(param_name);
-      ORT_ENFORCE(it != named_parameters_.end(), "Unknown param: ", param_name, " for gradient input: ", input_name);
-      // TODO: don't pre-allocate the gradient buffer.
-      ORT_THROW_IF_ERROR(it->second->AllocateGrad(input_name, train_sess_state));
-      gradients_.push_back(it->second->Gradient());
     } else {
       user_input_names.emplace_back(input_name);
     }
   }
+
+  gradients_.resize(grad_input_names.size());
+
   train_input_names_ = user_input_names;
   train_input_names_.insert(train_input_names_.end(), param_input_names.begin(), param_input_names.end());
   train_input_names_.insert(train_input_names_.end(), grad_input_names.begin(), grad_input_names.end());
   train_input_names_.insert(train_input_names_.end(), reset_grad_name.begin(), reset_grad_name.end());
 
-  // Eval model validation
-  if (nullptr != eval_sess_) {
+  // Loop each parameter, allocate it's memory based on user specified device.
+  for (auto& param_name : param_input_names) {
+    ORT_ENFORCE(named_parameters.find(param_name) != named_parameters.end());
+    OrtValue& source_ortvalue = named_parameters[param_name]->Data();
+    ORT_ENFORCE(source_ortvalue.IsTensor());
+    const Tensor& source_tensor = source_ortvalue.Get<Tensor>();
+
+    std::vector<SessionState::NodeInfo> node_info_vec;
+    ORT_THROW_IF_ERROR(train_sess_state.GetInputNodeInfo(param_name, node_info_vec));
+    const auto& node_info = node_info_vec.front();
+    const auto target_device = *node_info.device;
+    for (auto it = node_info_vec.begin(); it != node_info_vec.end(); ++it) {
+      ORT_ENFORCE(target_device == *(it->device));
+    }
+
+    // Create parameter value copy with corresponding device user sets the session on.
+    // We did not re-use the data even CPU tensor is needed.
+    // TODO(pengwa): consider whether we should alloc contiguous buffer for parameters or gradients.
+    OrtValue target_ortvalue;
+    auto allocator = train_sess_state.GetAllocator(target_device);
+    ORT_ENFORCE(allocator != nullptr);
+
+    Tensor::InitOrtValue(source_tensor.DataType(),
+                         source_tensor.Shape(),
+                         allocator, target_ortvalue);
+    Tensor* target_tensor_ptr = target_ortvalue.GetMutable<Tensor>();
+    ORT_THROW_IF_ERROR(train_sess_state.GetDataTransferMgr().CopyTensor(source_tensor, *target_tensor_ptr));
+
+    auto param_share_ptr = std::make_shared<Parameter>(param_name, target_ortvalue, named_parameters[param_name]->RequiresGrad());
```

Comment:
so we are creating a new sharedptr to Parameter here for internal use of Module. The problem with this is that the optimizer will still use the older user-passed Parameter object, so they won't really be sharing the same object.

Reply:
instead we can see if we can mutate the Data field of an existing shared Parameter object

Reply:
this will also make sure that when the Optimizer state is created, it is based on the device of this transferred parameter

Reply:
Alternatively, pass the named_parameters as a shared pointer unordered map (`std::shared_ptr<std::unordered_map<std::string, Parameter>>`) and update the value of the corresponding parameter.

Reply:
> so we are creating a new sharedptr to Parameter here for internal use of Module. The problem with this is that the optimizer will still use the older user-passed Parameter object, so they won't really be sharing the same object.

Use the API this way will not have the issue you mention, right? Correct me if I understand it wrong. 

```
  LoadCheckpoint(checkpoint_to_load_path, state)
  auto model = std::make_unique<Module>(model_uri, state.module_checkpoint_state.named_parameters,);
  auto optimizer = Optimizer(optim_uri, model->NamedParameters(), session_option, *env);
```